beacon-chain/light-client-network.md

```diff
@@ -0,0 +1,182 @@
+
+```
+custom_data = Container(data_radius: uint256)
+custom_payload = serialize(custom_data)
+```
+
+### Routing Table
+
+The Beacon CHain Light Client Network uses the standard routing table structure from the Portal Wire Protocol.
+
+### Node State
+
+#### Data Radius
+
+The Beacon Chain Light Client Network includes one additional piece of node state that should be tracked. Nodes must track the `data_radius`
+from the Ping and Pong messages for other nodes in the network. This value is a 256 bit integer and represents the data that
+a node is "interested" in.
+
+We define the following function to determine whether node in the network should be interested in a piece of content:
+
+```
+interested(node, content) = distance(node.id, content.id) <= node.radius
+```
+
+A node is expected to maintain `radius` information for each node in its local node table. A node's `radius` value may fluctuate as the contents of its local key-value store change.
+
+A node should track their own radius value and provide this value in all Ping or Pong messages it sends to other nodes.
+
+### Data Types
+
+The beacon chain light client DHT stores the following data items:
+
+* LightClientBootstrap
+* LightClientUpdate
+
+The following data objects are ephemeral and we store only the latest values:
+
+* LightClientFinalityUpdate
+* LightClientOptimisticUpdate
+
+#### Constants
+
+We define the following constants which are used in the various data type definitions:
+
+```py
+# Maximum number of `LightClientUpdate` instances in a single request
+MAX_REQUEST_LIGHT_CLIENT_UPDATES = 2**7  # = 128
+```
+
+#### ForkDigest
+4-byte fork digest for the current beacon chain version and ``genesis_validators_root``.
+
+#### LightClientBootstrap
+
+```
+light_client_bootstrap_key = Container(block_hash: Bytes32)
+selector                   = 0x00
+
+content                    = ForkDigest || SSZ.serialize(LightlientBootstrap)
+content_key                = selector + SSZ.serialize(light_client_bootstrap_key)
+```
+
+#### LightClientUpdatesByRange
+
+```
+light_client_update_keys   = Container(start_period: uint64, count: uint64)
+selector                   = 0x01
+
+content                    = SSZList(ForkDigest || LightClientUpdate, max_lenght=MAX_REQUEST_LIGHT_CLIENT_UPDATES)
+content_key                = selector + SSZ.serialize(light_client_update_keys)
+```
+
+#### LightClientFinalityUpdate
+
+```
+light_client_finality_update_key  = Container(None)
+selector                          = 0x02
+
+content                           = ForkDigest || SSZ.serialize(light_client_finality_update)
+content_key                       = selector + SSZ.serialize(light_client_finality_update_key)
+```
+
+> A `None` in the content key is equivalent to the request for the latest
+LightClientFinalityUpdate that the requested node has available.
+
+#### LightClientOptimisticUpdate
+
+```
+light_client_optimistic_update_key   = Container(None)
```

Comment:
Conteiners without fields not allowed in ssz  - https://github.com/ethereum/consensus-specs/blob/dev/ssz/simple-serialize.md#illegal-types.

Maybe,  empty list -`List[byte, 0]`  or `Container(latest: bool)` ?  (Although the second form suggest there is request for not latest update)

Reply:
Or, while that might sound confusing, just empty.

It would be the equivalent of `Union[None]` in case you work with SSZ Union for the selectors.

Reply:
I think that we should be a little careful here.

First, if this results in a static `content-id` and this object is being regularly fetched from nodes in the network, then we end up with a hot-spot where nodes that live close to that spot will shoulder most of the responsibility for serving the data.

Second, and likely minor but maybe worth mentioning, if we use any "short" value here, we set ourselves up with a situation where we have to be careful not to map any other "short" values to the same location.

One option to solve the hot-spot issue would be to have the key change location on an epoch boundary every N seconds, picking a value of N that doesn't move things too fast or too slow based on what we expect access patterns to look like.

Reply:
The `LightClientFinalityUpdate` and `LightClientOptimisticUpdate` are ephemeral objects. The libp2p interface only allows for retrieval of the `latest` of these objects. And they also gets gossiped.
So I think the idea is that for the Portal network we do the same and only provide the latest.

With that in mind, I think it can function similarly like we were having the MasterAccumulator for the history network, before it became a static object added at compile time. That is, every node stores them / has them available and additionally, they get gossiped to all. This does require some changes to they way of requesting them and of gossiping.


> One option to solve the hot-spot issue would be to have the key change location on an epoch boundary every N seconds, picking a value of N that doesn't move things too fast or too slow based on what we expect access patterns to look like.

That could be an alternative to letting every node store this.
But on the gossip side, it would probably be better to gossip to all as I think that would make it easier for a node needs to follow the updates.

Reply:
I did just realize that the Union None value should be only on the 0 selector, and also not for two types: https://github.com/ethereum/consensus-specs/blob/dev/ssz/simple-serialize.md#union

Reply:
Changed this field to 0 (uint8).